File: src/SproAdapter/adoclientpeer.cs
using System;
using System.Data;

namespace SocketProAdapter.ServerSide
{
    public class CAdoClientPeer : CClientPeer
    {
        public delegate void DAdonetLoaded(CAdoClientPeer sender, ushort reqId);
        public event DAdonetLoaded OnAdonetLoaded;

        public CAdoSerializationHelper AdoSerializer
        {
            get
            {
                return m_AdoSerializer;
            }
        }

        [RequestAttr(CAdoSerializationHelper.idDataSetHeaderArrive)]
        private void LoadDataSetHeader()
        {
            m_AdoSerializer.Load(CAdoSerializationHelper.idDataSetHeaderArrive, m_qBuffer);
            if (OnAdonetLoaded != null)
                OnAdonetLoaded.Invoke(this, CAdoSerializationHelper.idDataSetHeaderArrive);
        }

        [RequestAttr(CAdoSerializationHelper.idDataTableHeaderArrive)]
        private void LoadDataTableHeader()
        {
            m_AdoSerializer.Load(CAdoSerializationHelper.idDataTableHeaderArrive, m_qBuffer);
            if (OnAdonetLoaded != null)
                OnAdonetLoaded.Invoke(this, CAdoSerializationHelper.idDataTableHeaderArrive);
        }

        [RequestAttr(CAdoSerializationHelper.idDataReaderHeaderArrive)]
        private void LoadDataReaderHeader()
        {
            m_AdoSerializer.Load(CAdoSerializationHelper.idDataReaderHeaderArrive, m_qBuffer);
            if (OnAdonetLoaded != null)
                OnAdonetLoaded.Invoke(this, CAdoSerializationHelper.idDataReaderHeaderArrive);
        }

        [RequestAttr(CAdoSerializationHelper.idEndDataSet)]
        private void LoadEndDataSet()
        {
            m_AdoSerializer.Load(CAdoSerializationHelper.idEndDataSet, m_qBuffer);
            if (OnAdonetLoaded != null)
                OnAdonetLoaded.Invoke(this, CAdoSerializationHelper.idEndDataSet);
        }

        [RequestAttr(CAdoSerializationHelper.idEndDataTable)]
        private void LoadEndDataTable()
        {
            m_AdoSerializer.Load(CAdoSerializationHelper.idEndDataTable, m_qBuffer);
            if (OnAdonetLoaded != null)
                OnAdonetLoaded.Invoke(this, CAdoSerializationHelper.idEndDataTable);
        }

        [RequestAttr(CAdoSerializationHelper.idEndDataReader)]
        private void LoadEndDataReader()
        {
            m_AdoSerializer.Load(CAdoSerializationHelper.idEndDataReader, m_qBuffer);
            if (OnAdonetLoaded != null)
                OnAdonetLoaded.Invoke(this, CAdoSerializationHelper.idEndDataReader);
        }

        [RequestAttr(CAdoSerializationHelper.idDataTableRowsArrive, true)]
        private void LoadDataTableRows()
        {
            m_AdoSerializer.Load(CAdoSerializationHelper.idDataTableRowsArrive, m_qBuffer);
            if (OnAdonetLoaded != null)
                OnAdonetLoaded.Invoke(this, CAdoSerializationHelper.idDataTableRowsArrive);
        }

        [RequestAttr(CAdoSerializationHelper.idDataReaderRecordsArrive, true)]
        private void LoadDataReaderRecords()
        {
            m_AdoSerializer.Load(CAdoSerializationHelper.idDataReaderRecordsArrive, m_qBuffer);
            if (OnAdonetLoaded != null)
                OnAdonetLoaded.Invoke(this, CAdoSerializationHelper.idDataReaderRecordsArrive);
        }

        public ulong Send(IDataReader dr)
        {
            return Send(dr, CAdoSerializationHelper.DEFAULT_BATCH_SIZE);
        }

        public virtual ulong Send(IDataReader dr, uint batchSize)
        {
            uint res;
            ulong nSize = 0;
            bool bSuc;
            if (dr == null)
                throw new ArgumentException("Must pass in a valid data reader interface!");
            using (CScopeUQueue su = new CScopeUQueue())
            {
                CUQueue UQueue = su.UQueue;
                bool bBatching = Batching;
                if (!bBatching)
                    bSuc = StartBatching();
                do
                {
                    UQueue.SetSize(0);
                    m_AdoSerializer.PushHeader(UQueue, dr);
                    if (batchSize < 2048)
                        batchSize = 2048;
                    UQueue.Save(batchSize);
                    nSize = res = SendResult(CAdoSerializationHelper.idDataReaderHeaderArrive, UQueue);
                    UQueue.SetSize(0);

                    //monitor socket close event and cancel request
                    if (res == CClientPeer.REQUEST_CANCELED || res == CClientPeer.SOCKET_NOT_FOUND)
                        break;

                    while (dr.Read())
                    {
                        m_AdoSerializer.Push(UQueue, dr);
                        if (UQueue.GetSize() > batchSize)
                        {
                            res = SendResult(CAdoSerializationHelper.idDataReaderRecordsArrive, UQueue);
                            UQueue.SetSize(0);

                            //monitor socket close event and cancel request
                            if (res == CClientPeer.REQUEST_CANCELED || res == CClientPeer.SOCKET_NOT_FOUND)
                            {
                                nSize = res;
                                break;
                            }
                            else
                            {
                                nSize += res;
                                if (BytesBatched > 2 * batchSize)
                                {
                                    //if we find too much are stored in batch queue, we send them and start a new batching
                                    bSuc = CommitBatching();
                                    bSuc = StartBatching();
                                }
                            }
                        }
                    }
                    if (UQueue.GetSize() > 0) //remaining
                    {
                        res = SendResult(CAdoSerializationHelper.idDataReaderRecordsArrive, UQueue);
                        UQueue.SetSize(0);

                        //monitor socket close event and cancel request
                        if (res == CClientPeer.REQUEST_CANCELED || res == CClientPeer.SOCKET_NOT_FOUND)
                        {
                            nSize = res;
                            break;
                        }
                        nSize += res;
                    }
                } while (false);
                UQueue.SetSize(0);
                res = SendResult(CAdoSerializationHelper.idEndDataReader);
                //monitor socket close event and cancel request
                if (res == CClientPeer.REQUEST_CANCELED || res == CClientPeer.SOCKET_NOT_FOUND)
                    nSize = res;
                else
                    nSize += res;

                if (!bBatching)
                    bSuc = CommitBatching();
            }
            return nSize;
        }

        public virtual ulong Send(DataSet ds, bool bNeedRelations, uint batchSize)
        {
            bool bSuc;
            uint res;
            ulong nSize = 0;
            if (ds == null)
                throw new ArgumentException("Must pass in an valid DataSet object!");
            using (CScopeUQueue su = new CScopeUQueue())
            {
                CUQueue UQueue = su.UQueue;
                m_AdoSerializer.PushHeader(UQueue, ds);
                bool bBatching = Batching;
                if (!bBatching)
                    bSuc = StartBatching();
                do
                {
                    nSize = res = SendResult(CAdoSerializationHelper.idDataSetHeaderArrive, UQueue);
                    UQueue.SetSize(0);

                    //monitor socket close event and cancel request
                    if (res == CClientPeer.REQUEST_CANCELED || res == CClientPeer.SOCKET_NOT_FOUND)
                        break;
                    foreach (DataTable dt in ds.Tables)
                    {
                        ulong rtn = Send(dt, batchSize);
                        UQueue.SetSize(0);

                        if (rtn == CClientPeer.REQUEST_CANCELED || rtn == CClientPeer.SOCKET_NOT_FOUND)
                        {
                            nSize = rtn;
                            break;
                        }
                        else
                            nSize += rtn;
                    }

                    res = EndDataSet(ds, bNeedRelations, UQueue);
                    if (res == CClientPeer.REQUEST_CANCELED || res == CClientPeer.SOCKET_NOT_FOUND)
                    {
                        nSize = res;
                        break;
                    }
                    else
                        nSize += res;
                } while (false);
                UQueue.SetSize(0);
                if (!bBatching && Batching)
                    bSuc = CommitBatching();
            }
            return nSize;
        }

        uint EndDataSet(DataSet ds, bool bNeedRelations, CUQueue UQueue)
        {
            UQueue.SetSize(0);
            if (bNeedRelations && ds.Relations != null)
                m_AdoSerializer.Push(UQueue, ds.Relations);
            return SendResult(CAdoSerializationHelper.idEndDataSet, UQueue);
        }

        public ulong Send(DataSet ds, bool bNeedRelations)
        {
            return Send(ds, bNeedRelations, CAdoSerializationHelper.DEFAULT_BATCH_SIZE);
        }

        public ulong Send(DataSet ds)
        {
            return Send(ds, false, CAdoSerializationHelper.DEFAULT_BATCH_SIZE);
        }

        public ulong Send(DataTable dt)
        {
            return Send(dt, CAdoSerializationHelper.DEFAULT_BATCH_SIZE);
        }

        public virtual ulong Send(DataTable dt, uint batchSize)
        {
            uint res;
            ulong nSize;
            bool bSuc;
            if (dt == null)
                throw new ArgumentException("Must pass in an valid DataTable object!");
            using (CScopeUQueue su = new CScopeUQueue())
            {
                CUQueue UQueue = su.UQueue;
                bool bBatching = Batching;
                if (!bBatching)
                    bSuc = StartBatching();
                do
                {
                    //m_AdoSerializer->PushHeader(UQueue, dt, bNeedParentRelations, bNeedChildRelations);
                    m_AdoSerializer.PushHeader(UQueue, dt, false, false);
                    if (batchSize < 2048)
                        batchSize = 2048;
                    UQueue.Save(batchSize);
                    nSize = res = SendResult(CAdoSerializationHelper.idDataTableHeaderArrive, UQueue);
                    UQueue.SetSize(0);

                    //monitor socket close event and cancel request
                    if (res == CClientPeer.REQUEST_CANCELED || res == CClientPeer.SOCKET_NOT_FOUND)
                        break;

                    foreach (DataRow dr in dt.Rows)
                    {
                        m_AdoSerializer.Push(UQueue, dr);
                        if (UQueue.GetSize() > batchSize)
                        {
                            res = SendResult(CAdoSerializationHelper.idDataTableRowsArrive, UQueue);
                            UQueue.SetSize(0);

                            //monitor socket close event and cancel request
                            if (res == CClientPeer.REQUEST_CANCELED || res == CClientPeer.SOCKET_NOT_FOUND)
                            {
                                nSize = res;
                                break;
                            }
                            else
                            {
                                if (BytesBatched > 2 * batchSize)
                                {
                                    bSuc = CommitBatching();
                                    bSuc = StartBatching();
                                }
                                nSize += res;
                            }
                        }
                    }
                    if (res == CClientPeer.REQUEST_CANCELED || res == CClientPeer.SOCKET_NOT_FOUND)
                        break;
                    if (UQueue.GetSize() > 0) //remaining
                    {
                        res = SendResult(CAdoSerializationHelper.idDataTableRowsArrive, UQueue);
                        UQueue.SetSize(0);

                        //monitor socket close event and cancel request
                        if (res == CClientPeer.REQUEST_CANCELED || res == CClientPeer.SOCKET_NOT_FOUND)
                        {
                            nSize = res;
                            break;
                        }
                        else
                            nSize += res;
                    }
                } while (false);
                UQueue.SetSize(0);
                res = SendResult(CAdoSerializationHelper.idEndDataTable);
                if (res == CClientPeer.REQUEST_CANCELED || res == CClientPeer.SOCKET_NOT_FOUND)
                    nSize = res;
                else
                    nSize += res;
                if (!bBatching)
                    bSuc = CommitBatching();
            }
            return nSize;
        }
        private CAdoSerializationHelper m_AdoSerializer = new CAdoSerializationHelper();
    }
}

File: SproAdapters/common/AdoClientPeer.cpp
#include "stdafx.h"
#include "AdoClientPeer.h"

using namespace System;
using namespace System::Data;
namespace SocketProAdapter
{
	namespace ServerSide
	{
		CAdoClientPeer::CAdoClientPeer()
		{
			m_AdoSerialier = gcnew CAsyncAdoSerializationHelper();
		}

		CAdoClientPeer::~CAdoClientPeer()
		{
			delete m_AdoSerialier;
		}

		void CAdoClientPeer::OnFastRequestArrive(short sRequestID, int nLen)
		{
			switch(sRequestID)
			{
			case CAsyncAdoSerializationHelper::idDataSetHeaderArrive:
			case CAsyncAdoSerializationHelper::idDataReaderHeaderArrive:
			case CAsyncAdoSerializationHelper::idDataTableHeaderArrive:
			case CAsyncAdoSerializationHelper::idDataTableRowsArrive:
			case CAsyncAdoSerializationHelper::idDataReaderRecordsArrive:
			case CAsyncAdoSerializationHelper::idEndDataReader:
			case CAsyncAdoSerializationHelper::idEndDataTable:
			case CAsyncAdoSerializationHelper::idEndDataSet:
				if(!m_bSlowSet)
				{
					CBaseService ^p = GetBaseService();
					bool bSuc = p->AddSlowRequest(CAsyncAdoSerializationHelper::idDataReaderRecordsArrive);
					bSuc = p->AddSlowRequest(CAsyncAdoSerializationHelper::idDataTableRowsArrive);
					m_bSlowSet = true;
				}
				m_AdoSerialier->Load(sRequestID, m_UQueue);
				SendResult(sRequestID); //empty result
				break;
			default:
				break;
			}
		}

		int CAdoClientPeer::OnSlowRequestArrive(short sRequestID, int nLen)
		{
			switch(sRequestID)
			{
			case CAsyncAdoSerializationHelper::idDataSetHeaderArrive:
			case CAsyncAdoSerializationHelper::idDataReaderHeaderArrive:
			case CAsyncAdoSerializationHelper::idDataTableHeaderArrive:
			case CAsyncAdoSerializationHelper::idDataTableRowsArrive:
			case CAsyncAdoSerializationHelper::idDataReaderRecordsArrive:
			case CAsyncAdoSerializationHelper::idEndDataReader:
			case CAsyncAdoSerializationHelper::idEndDataTable:
			case CAsyncAdoSerializationHelper::idEndDataSet:
				m_AdoSerialier->Load(sRequestID, m_UQueue);
				SendResult(sRequestID); //empty result
				break;
			default:
				break;
			}
			return 0;
		}

		System::Int64 CAdoClientPeer::Send(IDataReader ^dr)
		{
			return Send(dr, 10240);
		}

		System::Int64 CAdoClientPeer::Send(IDataReader ^dr, int nBatchSize)
		{
			int res;
			System::Int64 nSize;
			bool bSuc;
			if (dr == nullptr)
				throw gcnew Exception("Must pass in a valid data reader interface!");
			CScopeUQueue su;
			CUQueue ^UQueue = su.UQueue;
			bool bBatching = IsBatching;
			if (!bBatching)
				bSuc = StartBatching();
			do
			{
				UQueue->SetSize(0);
				if(TransferServerException)
					UQueue->Push((long)0); //required by client
				m_AdoSerialier->PushHeader(UQueue, dr);
				if (nBatchSize < 2048)
					nBatchSize = 2048;
				UQueue->Push(nBatchSize);
				nSize = res = SendReturnData(CAsyncAdoSerializationHelper::idDataReaderHeaderArrive, UQueue);
				UQueue->SetSize(0);
				if(TransferServerException)
					UQueue->Push((long)0); //required by client
				//monitor socket close event and cancel request
				if (res == CClientPeer::REQUEST_CANCELED || res == CClientPeer::SOCKET_NOT_FOUND)
				{
					break;
				}
				
				while (dr->Read())
				{
					m_AdoSerialier->Push(UQueue, dr);
					if (UQueue->GetSize() > nBatchSize)
					{
						res = SendReturnData(CAsyncAdoSerializationHelper::idDataReaderRecordsArrive, UQueue);
						UQueue->SetSize(0);
						if(TransferServerException)
							UQueue->Push((long)0); //required by client
						//monitor socket close event and cancel request
						if (res == CClientPeer::REQUEST_CANCELED || res == CClientPeer::SOCKET_NOT_FOUND)
						{
							nSize = res;
							break;
						}
						else
						{
							nSize += res;
							if (BytesBatched > 2 * nBatchSize)
							{
								//if we find too much are stored in batch queue, we send them and start a new batching
								bSuc = CommitBatching();
								bSuc = StartBatching();
							}
						}
					}
				}
				if (UQueue->GetSize() > sizeof(long)) //remaining
				{
					res = SendReturnData(CAsyncAdoSerializationHelper::idDataReaderRecordsArrive, UQueue);
					UQueue->SetSize(0);
					if(TransferServerException)
						UQueue->Push((long)0); //required by client
					//monitor socket close event and cancel request
					if (res == CClientPeer::REQUEST_CANCELED || res == CClientPeer::SOCKET_NOT_FOUND)
					{
						nSize = res;
						break;
					}
					nSize += res;
				}
			}while(false);
			UQueue->SetSize(0);
			SendResult(CAsyncAdoSerializationHelper::idEndDataReader);
			if (!bBatching)
				bSuc = CommitBatching();
			return nSize;
		}

		System::Int64 CAdoClientPeer::Send(DataTable ^dt)
		{
			return Send(dt, 10240);
		}

		System::Int64 CAdoClientPeer::Send(DataTable ^dt, int nBatchSize)
		{
			int res;
			System::Int64 nSize;
			bool bSuc;
			if (dt == nullptr)
				throw gcnew Exception("Must pass in an valid DataTable object!");
			CScopeUQueue su;
			CUQueue ^UQueue = su.UQueue;
			if(TransferServerException)
				UQueue->Push((long)0); //required by client
			bool bBatching = IsBatching;
			if (!bBatching)
				bSuc = StartBatching();
			do
			{
				//m_AdoSerialier->PushHeader(UQueue, dt, bNeedParentRelations, bNeedChildRelations);
				m_AdoSerialier->PushHeader(UQueue, dt, false, false);
				if (nBatchSize < 2048)
					nBatchSize = 2048;
				UQueue->Push(nBatchSize);
				nSize = res = SendReturnData(CAsyncAdoSerializationHelper::idDataTableHeaderArrive, UQueue);
				UQueue->SetSize(0);
				if(TransferServerException)
					UQueue->Push((long)0); //required by client

				//monitor socket close event and cancel request
				if (res == CClientPeer::REQUEST_CANCELED || res == CClientPeer::SOCKET_NOT_FOUND)
					break;
				for each (DataRow ^dr in dt->Rows)
				{
					m_AdoSerialier->Push(UQueue, dr);
					if (UQueue->GetSize() > nBatchSize)
					{
						res = SendReturnData(CAsyncAdoSerializationHelper::idDataTableRowsArrive, UQueue);
						UQueue->SetSize(0);
						if(TransferServerException)
							UQueue->Push((long)0); //required by client
						//monitor socket close event and cancel request
						if (res == CClientPeer::REQUEST_CANCELED || res == CClientPeer::SOCKET_NOT_FOUND)
						{
							nSize = res;
							break;
						}
						else
						{
							if(BytesBatched > 2*nBatchSize)
							{
								bSuc = CommitBatching(); 
								bSuc = StartBatching();
							}
							nSize += res;
						}
					}
				}
				if (res == CClientPeer::REQUEST_CANCELED || res == CClientPeer::SOCKET_NOT_FOUND)
					break;
				if (UQueue->GetSize() > sizeof(long)) //remaining
				{
					res = SendReturnData(CAsyncAdoSerializationHelper::idDataTableRowsArrive, UQueue);
					UQueue->SetSize(0);
					if(TransferServerException)
						UQueue->Push((long)0); //required by client
					//monitor socket close event and cancel request
					if (res == CClientPeer::REQUEST_CANCELED || res == CClientPeer::SOCKET_NOT_FOUND)
					{
						nSize = res;
						break;
					}
					nSize += res;
				}
			}while(false);
			UQueue->SetSize(0);
			SendResult(CAsyncAdoSerializationHelper::idEndDataTable);
			if (!bBatching)
				bSuc = CommitBatching();
			return nSize;
		}

		System::Int64 CAdoClientPeer::Send(DataSet ^ds, bool bNeedRelations)
		{
			return Send(ds, bNeedRelations, 10240);
		}

		System::Int64 CAdoClientPeer::Send(DataSet ^ds, bool bNeedRelations, int nBatchSize)
		{
			bool bSuc;
			int res;
			if (ds == nullptr)
				throw gcnew Exception("Must pass in an valid DataSet object!");
			CScopeUQueue su;
			CUQueue ^UQueue = su.UQueue;
			System::Int64 nSize = 0;
			if(TransferServerException)
				UQueue->Push((long)0); //required by client
			m_AdoSerialier->PushHeader(UQueue, ds);
			bool bBatching = IsBatching;
			if (!bBatching)
				bSuc = StartBatching();
			do
			{
				nSize = res = SendReturnData(CAsyncAdoSerializationHelper::idDataSetHeaderArrive, UQueue);
				UQueue->SetSize(0);
				if(TransferServerException)
					UQueue->Push((long)0); //required by client
				//monitor socket close event and cancel request
				if (res == CClientPeer::REQUEST_CANCELED || res == CClientPeer::SOCKET_NOT_FOUND)
				{
					break;
				}
				for each (DataTable ^dt in ds->Tables)
				{
					System::Int64 rtn = Send(dt, nBatchSize);
					UQueue->SetSize(0);
					if(TransferServerException)
						UQueue->Push((long)0); //required by client
					if (rtn == CClientPeer::REQUEST_CANCELED || rtn == CClientPeer::SOCKET_NOT_FOUND)
					{
						nSize = rtn;
						break;
					}
					nSize += rtn;
				}
				if (res == CClientPeer::REQUEST_CANCELED || res == CClientPeer::SOCKET_NOT_FOUND)
				{
					break;
				}
				res = EndDataSet(ds, bNeedRelations, UQueue);
				if (res == CClientPeer::REQUEST_CANCELED || res == CClientPeer::SOCKET_NOT_FOUND)
				{
					nSize = res;
					break;
				}
				nSize += res;
			}while(false);
			UQueue->SetSize(0);
			if (!bBatching && IsBatching)
				bSuc = CommitBatching();
			return nSize;
		}

		int CAdoClientPeer::EndDataSet(DataSet ^ds, bool bNeedRelations, CUQueue ^UQueue)
		{
			UQueue->SetSize(0);
			if(TransferServerException)
				UQueue->Push((long)0); //required by client
			if (bNeedRelations && ds->Relations != nullptr)
				m_AdoSerialier->Push(UQueue, ds->Relations);
			return SendReturnData(CAsyncAdoSerializationHelper::idEndDataSet, UQueue);
		}

		System::Int64 CAdoClientPeer::Send(DataSet ^ds)
		{
			return Send(ds, false, 10240);
		}
	}
}